### ISSUE TITLE
How to use the extension on a Pull Request

### ISSUE THREAD

**Issue by buckleyGI** [JoostVoskuil/azure-devops-gitleaks#15]
Hi,

We are trying to use the extension so that only secrets introduced by a pull request are scanned.
This to keep the main branch free of secrets.

Experimenting with the depth parameter did not scan all of the changes in the Pull Request.
The thinking was that setting it to 1 would only scan the merge commit that brings in the changes.

Since this is a DevOps extension, and I think it is a common use case, I am very willing to contribute to code/documentation that would others help out as well. Can you set me in the right direction? Thanks! 

This is what we tried so far

```
  - task: Gitleaks@1
    inputs:
      scanfolder: '$(Build.SourcesDirectory)'
      verbose: true
      depth: 1
```

```
  - task: Gitleaks@1
    inputs:
      scanfolder: '$(Build.SourcesDirectory)'
      verbose: true
      arguments: '--branch=$(System.PullRequest.SourceBranch)'
```

**Comment by JoostVoskuil** [comment]
Hi @buckleyGI,

Interesting thought! And that is I think a very good use case. Using the task only as build-validation because 'no junk on the trunk'. I really like that!

The depth option and setting it to 1 is not correct. With that option it only scans the latest commit. 

Currently the task is not branch aware; Gitleaks just scans al the commits despite the branch you are working from (so from the pipeline). Yes, it can be debated if that is correct and maybe the –branch flag should be provided by default. 
However, changing the default behavior of the task requires building a version 2 of the task because It might lead to surprises from users.

When providing the –branch arguments it scans all the commits of that branch, so including the commits that are done on the origin of the branch. 

So when I have
1.	Main has 200 commits
2.	Create branch from main named ‘test’ 
3.	Commit on test

Then Gitleaks scans for 201 commits on the test branch and 200 on the main branch when the branch arguments is provided.

If you want to scan only the commits done on step 3 then we need to provide a commit file. We can harvest that from the Azure DevOps API for a pull request: GET https://dev.azure.com/{organization}/{project}/_apis/git/repositories/{repositoryId}/pullRequests/{pullRequestId}/commits?api-version=6.1-preview.1

So, what I want to suggest is:

- By default, the task is branch unaware. This is something for v2 to consider (that needs also be compatible with GitLeaks 8 so the behavior is going to change)
- However, when the task is running in a ‘build-validation’ pipeline (can be detected when variable Build.Reason is ‘PullRequest’):
   - Fetch the commits from the Azure DevOps API through the API specified above and construct a commit-file (in the node library this is gitApi.getPullRequestCommits)
   - Append the –Branch=$(System.PullRequest.SourceBranch) attribute to Gitleaks
   - Append the –commit-file=$(file) attribute to GitLeaks

Is this something you want to try to build in/contribute to? If not that is also fine, I have some time to implement it.

What are your troughts?

**Comment by buckleyGI** [comment]
We have numerous quality checks when code wants to go to main/trunk. The "no junk on the trunk" :) will become just another quality check indeed.

> The depth option and setting it to 1 is not correct. With that option it only scans the latest commit.

Yep. Since this is a Pull Request that commit 1 will be the merge of the target and source branch.
Commits 2 and 3 will come from the target and source branch I presume but after that it becomes blurry which commits are processed.
I believe this isn't explicitly documented by Gitleaks and there is also no way to let it output the hashes of the commits it is processing.

Also, if we had that information, current commit and total commits, we could provide a progress bar (like the checkout taks does) by way of [this logging command](https://docs.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=bash#setprogress-show-percentage-completed.).

This would provide value especialy for scenario 2 (see below) where It taks more that 6 hours on our repo with the combined default toml and MS credential security adaption!
But that is off topic :) and a new improvement that has to come from the Gitleaks project.

The current behavior, no specific branch, is a good default IMO. It depends on your scenario right? (scenario 3)

These are the scenario's I think the extension will be commonly used for

1. scan the current state of a branch (so no history) as can be done with --no-git
2. scan the whole repo including history
3. scan a branch including history
4. scan commits before they reach the repo as can be done by a pre commit hook (out of scope for this extension but worth mentioning just for this reason)
5. scan commits that try to get from one branch to another e.g. a pull request from a feature branch to main ("no junk on the trunk")

Do you think it is exhaustive? I am planning to write some guidance on each scenario on the wiki.

I was also coming to the conclusion that for scenario 5 we had to do some customization. Gitleaks provides primitives to make integration with the different platform possible, but that's it.
For a while I thought there was something in the box that I missed but we have to give it indeed a file with the commits to scan for.

I am currently writing a solution that asks for the commits with a git command and I am aware you can use the REST Api for that as well.
The command I first came up with is something like this

`git diff main... | grep '^+ ' > c:\Added.txt`

But this looks better

`git rev-list main..`

The first git command wil only show added lines (hence the grep) while the second wil list all the commits that are new on the source branch.
The first command requires some more plumbing where you point Gitleaks to this 1 file only.
While the second command exactly gives you what Gitleaks expect, a newline separated list of commits.

This is WIP at my end bu I think the second command could give us what we need.

> can be detected when variable Build.Reason is ‘PullRequest’

Yep, that is how we currently do it

> Append the –commit-file=$(file) attribute to GitLeaks

Yep

> Append the –Branch=$(System.PullRequest.SourceBranch) attribute to Gitleaks

Is this still necessary? We have give Gitleaks the exact commits to scan for. Is it still required to pass a branch?

> Is this something you want to try to build in/contribute to? If not that is also fine, I have some time to implement it.

I will now implement it with some plumbing and may come round to integrating it.
My day job will probably slow down progress on this feature so if you can implement it that would be great.
I'll be your beta tester then :)

What I can do at this point is provide documentation/wiki from a newcomer perspective and write about how to implement the above common scenario's.

I'll let you know when we have a working solution in place.
The simple `git rev-list main.. `seems promising but still have to eat the pudding :)

**Comment by buckleyGI** [comment]
Today I worked on the 2 approaches: rev-list and diff

**rev-list**

_Working example_

- job: Gitleaks
  steps:
  - script: git rev-list origin/main.. > diff.txt
  - task: Gitleaks@1
    inputs:
      scanfolder: '$(Build.SourcesDirectory)'
      arguments: '--commits-file=diff.txt'

Remarks:
1. Gitleaks reports 2 times for the same secret: the merge commit and the original
2. adding a commit that removes the secret still triggers Gitleaks on reporting a secret for the original secret (of course)

Remark 2 makes this approach not usable IMO. This is comparable wtih a call to the REST API as you proposed earlier.

You could think that a way out of it would be to squash merge or do an interactive rebase on the source branch (cumbersome and demands a bit of knowledge).
But this isn't the solution either as Azure DevOps does a traditional merge commit before applying the branch policies on the PR and you can't change that.
The original commit with the secret will always be there and Gitleaks will always protest. Do you see a way out of it?

**diff**

_Working example_

- job: Gitleaks
  steps:
  - script: |
      git diff origin/main... | grep '^+' > diff.txt
      git add diff.txt
      git config --global user.email "youx@example.com"
      git config --global user.name "Yourx Name"
      git commit -m 'add diff.txt'
  - publish: diff.txt
    artifact: diff
  - task: Gitleaks@1
    inputs:
      scanfolder: '$(Build.SourcesDirectory)'
      depth: 1

Remarks
1. The diff.txt will contain what I call the "net effect" of all the commits: if a secret was added with one commit and removed with another it will no be present in the net effect.
It greps for all the additions (lines that start with a +), commits the file (required for Gitleaks to work but canbe considered a throwaway commit) and with a depth of 1 it will only scan the commit with this diff.txt
2. I tried the --path=diff.txt switch as the file to scan as that was more natural at first. I could not get it to work though.
`level=error msg="stat /home/vsts/work/1/s/diff.txt/.git: not a directory`
3. For debugging purposes I add the diff.txt as an artifact

So the diff approach is for me the winner ATM: It achieves the "no junk in the trunk" goal by checking the "net effect" to guarantee no secrets will be merged in.

**Comment by buckleyGI** [comment]
For reference this is the complete yaml. I'll gather some data for it next week and see if it meets expectations


```
trigger: none

schedules:
- cron: '0 0 * * SAT' # times are in UTC
  displayName: 'Gitleaks schedule'
  branches:
    include:
    - main
  always: true

jobs:

- job: Gitleaks
  steps:
  - script: |
      # Transform refs/heads/main to origin/main
      targetBranch="$(system.pullRequest.targetBranch)"
      prefixDevOps="origin"
      targetBranchOrigin="${targetBranch/refs\/heads/"$prefixDevOps"}"

      # Generate the "net effect" of all the commits
      git diff $targetBranchOrigin... | grep '^+' > GitleaksAdditions.txt
      git add GitleaksAdditions.txt

      # Add dummy/throwaway commit
      git config --global user.email "diff@dummy.com"
      git config --global user.name "diff bot"
      git commit -m 'GitleaksAdditions.txt'

    displayName: Generate additions
  - publish: GitleaksAdditions.txt
    displayName: Publish additions
    artifact: GitleaksAdditions
  - task: Gitleaks@1
    inputs:
      scanfolder: '$(Build.SourcesDirectory)'
      configtype: 'custom'
      configfile: 'gitleaks.toml'
      ${{ if eq(variables['Build.Reason'], 'PullRequest') }}:
        depth: 1 # Only scan the 1 commit that contains GitleaksAdditions.txt
      ${{ else }}:
        nogit: true # scan the repo as it currently is taking no history (commits) into account
      reportformat: 'sarif'
      redact: true
      version: '7.5.0' # At the time of writing (2021-11-10) there was an issue with later versions > 7.5.0 https://github.com/zricethezav/gitleaks/issues/630
      verbose: true


```

**Comment by JoostVoskuil** [comment]
1. The new version of the task has now pre-validation mode. If the build is run as part of a pull request, it fetches the commits from the pull requests and only scan those commits. You are right, selecting branch is not required.
2. I don't see how the merge options are important. Merging (with or without squash) is done after the pull request. So it is about getting the commits done prior to the merge. Maybe I don't understand you, can you ellaborate?
3. Please keep in mind that when a leak is found, the damage is already done. We only want to limit the number of commits to scan because of speed. The junk is already been made. When secrets are found you need to remove them with BFG. When running in pre-validation/pullrequest this means you cannot merge until this is resolved. 
4. Because of 3 I want to remove the --no-git option in v2. I don't see the value of it, same for scanning a particular branch. Yes there is value regarding Gitleaks but not in a pipeline imho
5. Progress bar is not possible, since this task is only a tool wrapper, not the tool itself.

**Comment by buckleyGI** [comment]
Hi Joost

We are pinning the task to v1.4.20 for the moment as the new feature that makes the task PR aware does not fit our use case.

Our use case is that we allow secrets in the history but not in the current state if you would do a get latest of main.
This corresponds to a clean secretless deploy of the files to production. 

This differs from your envisioned use case where every introduced secret should be nuked.
We allow the developer to make a correction/remove the secret in a next commit on the feature branch before it integrates to the main branch.

True, if a hacker gets the whole repo the secret is compromised.
Our concern is that a secret doesn't get deployed. I'll come back to that later in more detail if our use case is not clear.

2 Asks

1. Can you output the commits beign scanned? I saw it already does it! That's fast.
I think that even when I place myself in your use case, scan every new commit since divergence from  main, we noticed a bug.

The convention in DevOps is that if system.debug=true the task can output some more information for troubleshooting.

2. Can you make the new PR feature configurable so you can opt out? (commits-file)
This way our use case is supported as well and we can remove the pinning of the task and receive updates.

Thanks for the help! I'll come back to your latest reply a bit later.

**Comment by buckleyGI** [comment]
> 1. The new version of the task has now pre-validation mode. If the build is run as part of a pull request, it fetches the commits from the pull requests and only scan those commits. You are right, selecting branch is not required.

Thanks for the fast response. We already see it in action in our organisation. As said previously, we *might* have seen a bug here but can't see what commits are being scannend.

>     2. I don't see how the merge options are important. Merging (with or without squash) is done after the pull request. So it is about getting the commits done prior to the merge. Maybe I don't understand you, can you ellaborate?

If the explanation of our use case in the previous post was clear, we our only concerned of what gets merged into main.
(This is BTW not our end goal. We should review the 1000's of potential secrets in our coming sprint_s_ but for now we are only concerted about the influx of new secrets to main.)
Feature branches themselves in our organisation are allowed to have a commit with a secret that is later corrected in another commit in the same feature branch.
When going to main however, we don't allow those commits be merged in. A way to prevent that is with a squash right?

Once the feature branch is merged in, the branch is deleted and will be cleaned up after a while by a git maintenance that happens automatically by git once in a while (can also be forece).

>     3. Please keep in mind that when a leak is found, the damage is already done. We only want to limit the number of commits to scan because of speed. The junk is already been made. When secrets are found you need to remove them with BFG. When running in pre-validation/pullrequest this means you cannot merge until this is resolved.

This seems the good default but is a bit strict for certain use cases so can you make it optional/configurable? Our use case is that no secrets get into the main branch so they are not deployed to production.

>     4. Because of 3 I want to remove the --no-git option in v2. I don't see the value of it, same for scanning a particular branch.
>        Yes there is value regarding Gitleaks but not in a pipeline imho

We get real value out of --no-git as it scans what would go to production. In fact, we have a weekly pipeline that scans main with --no-git as new rules can be added/updates that weren't checked when the PR ran in the past. This would pick up any false negatives as new rules are added and is *required* IMO in addition to a quality check on the PR.

>     5. Progress bar is not possible, since this task is only a tool wrapper, not the tool itself.

Indeed, I'll make a request in the gitleaks repo. Some of our scans takes hours without feedback which would make some newcomers think something is wrong and abandon the extension.

**Comment by JoostVoskuil** [comment]
I have added a flag 'prevalidationbuild' to control the behaviour but thinking about removing that in v2 of the task. Please keep in mind that while you squash merge, the exposed secret can still be found in the pull request info/commits in Azure DevOps.

Also when system.debug mode is true, you see the scanned commits. 

For now I will close this issue. I am going to work on Gitleaks V8 support

**Comment by buckleyGI** [comment]
I hope we will have some time to update to V2
We are still a happy user and get great value out of the task
Thanks again @JoostVoskuil
